### PR TITLE
Use new evaluate in template plugin

### PR DESCRIPTION
**Pull request by praasz**
### Details:
 - Use new evaluate in template plugin as default. Add tensor conversion utils to make evaluate compatible. 
 - Continue to work on #13944 

### Tickets:
 - 80875

### CODE REVIEW

File: src/core/src/op/softsign.cpp [hunk 1]
```diff
@@ -90,6 +91,11 @@ bool ov::op::v9::SoftSign::evaluate(ov::TensorVector& outputs,
 
     const auto& in = inputs[0];
     auto& out = outputs[0];
-    out.set_shape(in.get_shape());
+
+    if (ov::util::is_dynamic_shape(out.get_shape())) {
+        out = Tensor(in.get_element_type(), in.get_shape());
+    } else {
+        out.set_shape(in.get_shape());
+    }
```
Comment (jane-intel) [inline src/core/src/op/softsign.cpp:99]: why did you change it only here? I see that lots of evaluate methods set shape to the output tensor. On the other hand I think that evaluate should not work with DynamicTensors at all.

Oh I see. This is the ov::Tensor evaluate -- I see lots of HostTensor vectors.
Anyway -- it looks strange that we must check for dynamic shape before setting new shape.
@ilyachur please advice what should we do?

Comment (ilyachur) [reply src/core/src/op/softsign.cpp:99]: New tensors don't support dynamic shapes. `{0, max_int}` is a WA for our internal code only in order to have a bridge between two APIs.
So I think you can just check the tensor size and call set_shape in case of empty tensor.

We don't need to use out internal API here. Maybe we need to modify `set_shape()` to support this WA

File: src/core/src/op/divide.cpp [hunk 2]
```diff
@@ -72,7 +73,7 @@ ov::Tensor equality_mask(const ov::Tensor& tensor, const shared_ptr<op::Constant
 }
 
 ov::Tensor or_tensor(const ov::Tensor& lhs, const ov::Tensor& rhs) {
-    auto outs = ov::TensorVector{{lhs.get_element_type(), Shape{0}}};
+    auto outs = ov::TensorVector{{lhs.get_element_type(), ov::util::make_dynamic_shape()}};
```
Comment (ilyachur) [inline src/core/src/op/divide.cpp:76]: Why do we need to use here the new WA? I would like to have it only as some internal temporary thing in order to connect two APIs.

File: src/core/dev_api/shape_util.hpp [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+namespace ov {
+namespace util {
+/**
+ * @brief Makes spacial version of 2D ov::Shape which is recognize as dynamic.
+ *
+ * This is special case used for tensor <-> host tensor conversion to indicate that tensor got dynamic shape.
+ *
+ * @return 2-D shape with {0, SIZE_MAX}
+ */
+OPENVINO_API Shape make_dynamic_shape();
```
Comment (ilyachur) [inline src/core/dev_api/shape_util.hpp:18]: Can we immediately mark these methods as deprecated in order to remove together with legacy API?

File: src/core/dev_api/tensor_conversion_util.hpp [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+namespace util {
+
+/**
+ * @brief Make temporay tensor from host tensor pointer.
+ *
+ * @param t  Input tensor for conversion.
+ * @return ov::Tensor which points to host tensor data. Can return not allocated or special dynamic depends on input
+ * tensor state.
+ */
+OPENVINO_API Tensor make_tmp_tensor(const ngraph::HostTensorPtr& t);
```
Comment (ilyachur) [inline src/core/dev_api/tensor_conversion_util.hpp:20]: Maybe `wrap_tensor`?
Also I propose to deprecate these methods.

**Comment by jane-intel** [comment]
@praasz please resolve conflicts

File: src/core/src/op/softsign.cpp (hunk 1 above)
Comment (praasz) [reply src/core/src/op/softsign.cpp]: Updated `set_shape` member function.

File: src/core/src/op/divide.cpp (hunk 2 above)
Comment (praasz) [reply src/core/src/op/divide.cpp]: Looks is not required here. Changed.

File: src/core/dev_api/shape_util.hpp (hunk 3 above)
Comment (praasz) [reply src/core/dev_api/shape_util.hpp]: Yes, marked and now is core internal only. Looks is not required by other libraries.

File: src/core/dev_api/tensor_conversion_util.hpp (hunk 4 above)
Comment (praasz) [reply src/core/dev_api/tensor_conversion_util.hpp]: Rename and marked deprecated.

**Review by jane-intel — APPROVED** [review]
LGTM

File: src/plugins/template/backend/int_executable.cpp [hunk 5]
```diff
@@ -130,21 +131,6 @@ bool runtime::interpreter::INTExecutable::call(const vector<shared_ptr<runtime::
             op_outputs.push_back(host_tensor);
         }
 
-        // get op type
```
Comment (iefode) [inline src/plugins/template/backend/int_executable.cpp:133]: As i remember, we made such w/a (according specs) to get correct results for Subgraph tests. Please make sure that these changes do not broke ov functional tests

File: src/plugins/template/backend/int_executable.cpp [hunk 6]
```diff
@@ -159,8 +145,13 @@ bool runtime::interpreter::INTExecutable::call(const vector<shared_ptr<runtime::
...
 
+        const auto tensor_inputs = ov::util::wrap_tensors(op_inputs);
+        auto tensor_outputs = ov::util::wrap_tensors(op_outputs);
+
         // Call evaluate for cloned_node with static shapes
-        if (!cloned_node->evaluate(op_outputs, op_inputs, eval_context)) {
+        if (cloned_node->evaluate(tensor_outputs, tensor_inputs, eval_context)) {
+            ov::util::update_output_host_tensors(op_outputs, tensor_outputs);
+        } else {
             evaluate_node(cloned_node, op_outputs, op_inputs);
```
Comment (iefode) [inline src/plugins/template/backend/int_executable.cpp:155]: what is `eval_context`? Would it affect [evaluate_map](https://github.com/openvinotoolkit/openvino/blob/91df0a8aa92dbcf67c85676e9a786cc360486b17/src/plugins/template/backend/evaluates_map.cpp#L4) to get reference results?

File: src/plugins/template/backend/int_executable.cpp (hunk 5 above)
Comment (praasz) [reply src/plugins/template/backend/int_executable.cpp:133]: In original implementation the `type` variable has been get from input but not used. For me it looks like remove not used code and should has no effect on functionality.

@artkuli Have you got any additional comment about this change?

File: src/plugins/template/backend/int_executable.cpp (hunk 6 above)
Comment (praasz) [reply src/plugins/template/backend/int_executable.cpp:155]: It looks like it is used by operators ReadValue, Assign. Current changes don't change the way how it is used.

@ilyachur 
Can you explain purpose of this `context`?

File: src/core/src/runtime/ov_tensor.cpp [hunk 7]
```diff
@@ -98,7 +99,14 @@ element::Type Tensor::get_element_type() const {
 }
 
 void Tensor::set_shape(const ov::Shape& shape) {
-    OV_TENSOR_STATEMENT(_impl->setShape({shape.begin(), shape.end()}));
+    // WA for tensor conversion from host tensor with dynamic shape.
```
Comment (ilya-lavrenov) [inline src/core/src/runtime/ov_tensor.cpp:102]: is it a temporary WA until remove legacy code?

Comment (ilyachur) [reply src/core/src/runtime/ov_tensor.cpp:102]: Yes it is. It is needed in order to support set_shape for tensors which were created from ngraph dynamic tensors

File: src/plugins/template/backend/int_executable.cpp (hunk 5 above)
Comment (ilyachur) [reply src/plugins/template/backend/int_executable.cpp:133]: CI should show if we have any issues here.

File: src/plugins/template/backend/int_executable.cpp (hunk 6 above)
Comment (ilyachur) [reply src/plugins/template/backend/int_executable.cpp:155]: Evaluation context is a runtime information which is needed for evaluate some operations. It is used in some memory ops like read value.

File: src/plugins/template/backend/int_executable.cpp (hunk 5 above)
Comment (iefode) [reply src/plugins/template/backend/int_executable.cpp:133]: @ilyachur @artkuli @praasz 
I meant 2 types of functional tests:
1. opset conformance
2. plugin functional

Anyway, I will check the changes to compare results with previous version

Unresolve discussion before my verification.
